go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -64,6 +66,249 @@ func NewVtctldServer(ts *topo.Server) *VtctldServer {
 	return &VtctldServer{ts: ts, tmc: tmclient.NewTabletManagerClient()}
 }
 
+// ChangeTabletType is part of the vtctlservicepb.VtctldServer interface.
+func (s *VtctldServer) ChangeTabletType(ctx context.Context, req *vtctldatapb.ChangeTabletTypeRequest) (*vtctldatapb.ChangeTabletTypeResponse, error) {
+	tablet, err := s.ts.GetTablet(ctx, req.TabletAlias)
+	if err != nil {
+		return nil, err
+	}
+
+	if !topo.IsTrivialTypeChange(tablet.Type, req.DbType) {
+		return nil, fmt.Errorf("tablet %v type change %v -> %v is not an allowed transition for ChangeTabletType", req.TabletAlias, tablet.Type, req.DbType)
+	}
+
+	if req.DryRun {
+		afterTablet := *tablet.Tablet
+		afterTablet.Type = req.DbType
+
+		return &vtctldatapb.ChangeTabletTypeResponse{
+			BeforeTablet: tablet.Tablet,
+			AfterTablet:  &afterTablet,
+			WasDryRun:    true,
+		}, nil
+	}
+
+	err = s.tmc.ChangeType(ctx, tablet.Tablet, req.DbType)
+	if err != nil {
+		return nil, err
+	}
+
+	changedTablet, _ := s.ts.GetTablet(ctx, req.TabletAlias)
+
+	return &vtctldatapb.ChangeTabletTypeResponse{
+		BeforeTablet: tablet.Tablet,
+		AfterTablet:  changedTablet.Tablet,
+		WasDryRun:    false,
+	}, nil
+}
+
+// CreateKeyspace is part of the vtctlservicepb.VtctldServer interface.
+func (s *VtctldServer) CreateKeyspace(ctx context.Context, req *vtctldatapb.CreateKeyspaceRequest) (*vtctldatapb.CreateKeyspaceResponse, error) {
```

Comment:
I understand these implementations are intended as a leg over from having the implementation tied up in `vtctl/vtctl.go::commandCreateKeyspace` etc but it feels odd that many of these implementations are so heavily duplicated and could result in drift. How do we plan on preventing that (rapid deprecation+removal of legacy impl is a totally valid answer).

Even so one thing I'd like us to consider is how we might want to detach the command functionality from the transport? E.g. what does it look like if we also wanted a REST API or something that kicked off create keyspace etc. (Viable answer here is : we only ever care about grpc but _if that changes_ the req/resp objects provide a clear way to introduce a layer of indirection so we don't have to duplicate code again).

Reply:
> but it feels odd that many of these implementations are so heavily duplicated and could result in drift. How do we plan on preventing that (rapid deprecation+removal of legacy impl is a totally valid answer).

Totally. I've stuck a [`VtctldServer` inside of the wrangler](https://github.com/vitessio/vitess/blob/master/go/vt/wrangler/wrangler.go#L48), which powers the legacy implementation. My plan here is to make another pass after this PR and replace all the logic with a call to `wr.VtctldServer().CreateKeyspace(...)` and so on, and then munge those responses back into the legacy output format. 

I didn't do this in this PR mainly because this branch was already gigantic, and it'll let me just focus on making sure I don't accidentally change the legacy output, which I've [already done once](https://github.com/vitessio/vitess/pull/7285).

> one thing I'd like us to consider is how we might want to detach the command functionality from the transport?

This one is really interesting. My answer for the legacy CLI is still to deprecate+remove. But, more abstractly about the REST case:

I'm not sure how we can guarantee client/server type safety without adding something like jsonschema to the mix, but ignoring that, I would probably use the grpc api as the de facto standard, and add adapters for REST/whatever, so we would:

1. Add a global flag to `vtctldclient` that says which transport to use. Add any extra flags needed to set up those transports (like the `grpc_static_auth_creds` flag for the grpc transport)
2. Create `package httpvtctldclient` in the same style as `package grpcvtctldclient`. This is where we'll translate the proto message objects into an HTTP request, and then the HTTP response back into a proto message. Similar for any other transports we want to support.
3. In `Root.PersistentPreRunE`, instantiate the correct client (`grpcvtctldclient.New(...)`, `httpvtctldclient.New(...)`, etc), everything about the individual command implementations is unchanged.